### PR TITLE
osd/scrub: use an AsyncReserver to handle scrub reservations on the replica side

### PR DESCRIPTION
**Pull request by ronen-fr**
That means that when trying to reserve its replicas for scrubbing, the primary sends a reservation request
with a 'queue this request' flag set. That request is queued at the scrub-reserver, and granted once
the number of concurrent 'remote reservations' falls below the configured threshold.

### CODE REVIEW

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
jenkins test make check

File: src/osd/scrubber/scrub_machine.h [hunk 1]
```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
...
+   * The 'failure' re 'log_failure' is logged if we are not reserved to
+   * begin with.
+   */
   void clear_reservation_by_remote_primary(bool log_failure);
 
+  /**
+   * cancel_ungranted_request - cancel the reservation request if it was not
+   * already granted by the async reserver.
+   */
+  void cancel_ungranted_request();
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.h:864]: cancel_ungranted_request and clear_reservation_by_remote_primary are very nearly the same thing.  I don't really understand why they are distinct.

File: src/osd/scrubber/scrub_machine.h [hunk 2]
```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
...
   /// handle a 'release' from a primary
   void on_release(const ReplicaRelease& ev);
 
-  /// cancel the reserver request.
-  /// The 'failure' re 'log_failure' is logged if we are not reserved to
-  /// begin with.
+  /**
+   * cancel the reserver request.
+   * The 'failure' re 'log_failure' is logged if we are not reserved to
+   * begin with.
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.h:856]: The "we are not reserved to begin with" is meant to be impossible, which is why you are logging it.  For cases that shouldn't be possible, either be *very* explicit or remove it entirely.

File: src/osd/scrubber/scrub_machine.h [hunk 3]
```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
...
+    void finish(int) override {}
+    void cancel() {
+      ceph_assert(pg->is_locked());
+      canceled = true;
+    }
+  };
+  // clang-format on
+
  private:
   bool reserved_by_my_primary{false};
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.h:912]: You didn't add this in this PR, but this is a really confusing name.  It's not possible to be reserved by anything other than the primary of the current interval.  The name could be interpreted to indicate either:
1. a reservation is active
2. that the active reservation is specifically for the primary

I'd really like the name changed to reservation_granted or something with a comment explaining that it is set to true iff we have sent a grant confirmation to the primary and the reservation has not yet been canceled either by the primary or by an interval change.

Comment (athanatos) [reply src/osd/scrubber/scrub_machine.h:912]: That would give us two state variables that fully express the current reservation state (pending_reservation_nonce and reservation_granted) and 3 possible states:
1. pending_reservation_nonce == 0 && !reservation_granted -- no reservation is requested or pending
2. pending_reservation_nonce != 0 && !reservation_granted -- we have a pending cb in the AsyncReserver for a request with nonce pending_reservation_nonce.
3. pending_reservation_nonce == 0 && reservation_granted -- we have sent a response to the primary granting the reservation

The syncronous/compatibility case bypasses 2.

Invariant: assert(!((pending_reservation_nonce != 0) && reservation_granted));

Whether you choose to adopt the above arrangement or not, I'd like to see a comment spelling out precisely what member variable combinations are possible.

**Review by athanatos — CHANGES_REQUESTED** [review]
Comments are mainly related to callback invalidation, pretty close.  PR structure was super easy to review, thanks!

File: src/osd/scrubber/scrub_machine.cc [hunk 4]
```diff
@@ -817,31 +817,42 @@ ReplicaReactCode ReplicaActive::on_reserve_request(
...
+    reserver_callback = new RtReservationCB(machine.m_pg, request_details);
+    reserver.request_reservation(pg_id, reserver_callback, 0, nullptr);
     next_action = ReplicaReactCode::goto_waiting_reservation;
 
   } else {
     // an immediate yes/no is required
-    const auto granted = svc.inc_scrubs_remote(scrbr->get_spgid().pgid);
+    reserver_callback = new NullOpCB(machine.m_pg);
+    const auto granted =
+	reserver.request_reservation_or_fail(pg_id, reserver_callback);
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.cc:838]: fixed the AsyncReserver

File: src/osd/scrubber/scrub_machine.cc [hunk 5]
```diff
@@ -817,31 +817,42 @@ ReplicaReactCode ReplicaActive::on_reserve_request(
...
+		    request_details)
+	     << dendl;
+    reserver_callback = new RtReservationCB(machine.m_pg, request_details);
+    reserver.request_reservation(pg_id, reserver_callback, 0, nullptr);
     next_action = ReplicaReactCode::goto_waiting_reservation;
 
   } else {
     // an immediate yes/no is required
-    const auto granted = svc.inc_scrubs_remote(scrbr->get_spgid().pgid);
+    reserver_callback = new NullOpCB(machine.m_pg);
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.cc:836]: Indeed... the actual reason was a different mistake.

File: src/osd/scrubber/scrub_machine.cc [hunk 6]
```diff
@@ -817,31 +817,42 @@ ReplicaReactCode ReplicaActive::on_reserve_request(
...
+  auto& reserver = m_osds->get_scrub_reserver();
 
   if (async_request) {
     // the request is to be handled asynchronously
-    svc.enqueue_remote_reservation(pg_id.pgid);
+    dout(20) << fmt::format(
+		    "{}: async request: {} details:{}", __func__, ev,
+		    request_details)
+	     << dendl;
+    reserver_callback = new RtReservationCB(machine.m_pg, request_details);
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.cc:830]: Agree.

File: src/osd/scrubber/scrub_machine.h (hunk 1 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.h:867]: cancel_ungranted_request() cancels just requests that were not granted, allowing
me to call it unconditionally at ReplicaWaitingReservation dtor.

**Comment by ronen-fr** [comment]
@athanatos - ready for re-review.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/scrubber/scrub_machine.cc [hunk 7]
```diff
@@ -800,45 +800,46 @@ ReplicaReactCode ReplicaActive::on_reserve_request(
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
   const auto m = ev.m_op->get_req<MOSDScrubReserve>();
-  const auto msg_nonce = m->reservation_nonce;
+  pending_reservation_nonce = m->reservation_nonce;
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/scrub_machine.cc:803]: Yes. A bug introduced by too many changes in a short time

File: src/osd/OSD.cc [hunk 8]
```diff
@@ -9889,6 +9891,9 @@ void OSD::handle_conf_change(const ConfigProxy& conf,
   if (changed.count("osd_max_trimming_pgs")) {
     service.snap_reserver.set_max(cct->_conf->osd_max_trimming_pgs);
   }
+  if (changed.count("osd_max_scrubs")) {
+    service.scrub_reserver.set_max(cct->_conf->osd_max_scrubs);
```
Comment (athanatos) [inline src/osd/OSD.cc:9895]: Nit: with this change we effectively allow osd_max_scrubs primary scrubs and osd_max_scrubs replica scrubs, right?  That's reasonable to me, but it's probably worth noting that in the config description.

File: src/osd/scrubber/scrub_machine.h [hunk 9]
```diff
@@ -776,22 +842,91 @@ struct ReplicaActive : sc::state<
...
+   * Note the interaction with 'pending_reservation_nonce': the combination
+   * of these two variables is used to track the state of the reservation
+   * with the scrub_reserver. The possible combinations:
+   * - pending_reservation_nonce == 0 && !reservation_granted -- no reservation
+   *    was granted, and none is pending;
+   * - pending_reservation_nonce != 0 && !reservation_granted -- we have a
+   *   pending cb in the AsyncReserver for a request with nonce
+   *   'pending_reservation_nonce'
+   * - pending_reservation_nonce == 0 && reservation_granted -- we have sent
+   *   a response to the primary granting the reservation
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.h:901]: Let's state the invariant explicitly:

```
 * invariant: !((pending_reservation_nonce != 0) && reservation_granted)
```

It's also worth noting how these states relate to the synchronous/compat case:

```
 * Note that in the event that the primary is too old to support asynchronous reservation,
 * MOSDScrubReserve::wait_for_resources will be set to false by the decoder and we
 * bypass the 2nd case above.  See ReplicaActive::on_reserve_request.
```

File: src/osd/scrubber/scrub_machine.h [hunk 10]
```diff
@@ -776,22 +842,91 @@ struct ReplicaActive : sc::state<
...
+   * a reservation request with this nonce is queued at the scrub_reserver,
+   * and was not yet granted.
+   */
+  MOSDScrubReserve::reservation_nonce_t pending_reservation_nonce{0};
+
+  // clang-format off
+  struct RtReservationCB : public Context {
+    PGRef pg;
+    AsyncScrubResData res_data;
+    bool canceled{false};
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.h:915]: No remaining users, right?  Let's remove it.

File: src/osd/scrubber_common.h [hunk 11]
```diff
@@ -24,6 +25,21 @@ namespace Scrub {
   struct ReplicaActive;
 }
 
+/// reservation-related data sent by the primary to the replicas,
+/// and used to match the responses to the requests
+struct AsyncScrubResData {
+  spg_t pgid;
+  pg_shard_t from;
+  epoch_t request_epoch;
+  MOSDScrubReserve::reservation_nonce_t nonce;
```
Comment (athanatos) [inline src/osd/scrubber_common.h:34]: Initialize nonce and request_epoch to 0 or disallow the default constructor and ensure that all constructors initialize them to a parameter.  We don't want it to be possible to construct an AsyncScrubResData with uninitialized fields.

File: src/osd/scrubber/scrub_machine.cc [hunk 12]
```diff
@@ -800,85 +795,118 @@ ReplicaReactCode ReplicaActive::on_reserve_request(
 {
   DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
   const auto m = ev.m_op->get_req<MOSDScrubReserve>();
-  const auto msg_nonce = m->reservation_nonce;
   dout(10) << fmt::format(
-		  "ReplicaActive::on_reserve_req() from {} request:{} is "
-		  "async?{} (reservation_nonce:{})",
-		  ev.m_from, ev, async_request, msg_nonce)
+		  "ReplicaActive::on_reserve_req() request:{} is async?{} "
```
Comment (athanatos) [inline src/osd/scrubber/scrub_machine.cc:799]: async_request rather than async in the log line.

**Review by athanatos — APPROVED** [review]
Remaining comments are straightforward, no need for another review once addressed.

File: src/osd/OSD.cc (hunk 8 above)
Comment (ronen-fr) [reply src/osd/OSD.cc:9895]: That change was introduced in another PR, on November 2023.
And - yes - I should change the documentation.

**Comment by ronen-fr** [comment]
merging based on my Teuthology runs

**Pull request merged**